### PR TITLE
Add timestamp to metadata

### PR DESCRIPTION
**Pull request by skiddykong**

### CODE REVIEW

File: aggregate-snapshot/aggregate-snapshot-service/src/test/java/uk/gov/justice/services/core/aggregate/SnapshotAwareAggregateServiceIT.java [hunk 1]
```diff
@@ -410,7 +413,14 @@ private DynamicallyLoadingClassLoader classLoaderWithGeneratedAggregateLoaded()
     private Stream<JsonEnvelope> createEventAndApply(long count, String eventName, TestAggregate aggregate) {
         List<Object> envelopes = new LinkedList<>();
         for (int i = 1; i <= count; i++) {
-            JsonEnvelope envelope = envelope().with(metadataWithRandomUUID(eventName).withStreamId(STREAM_ID)).withPayloadOf("value", "name").build();
+            JsonEnvelope envelope = DefaultJsonEnvelope
+                    .envelope()
```
Comment (purple52) [inline aggregate-snapshot/aggregate-snapshot-service/src/test/java/uk/gov/justice/services/core/aggregate/SnapshotAwareAggregateServiceIT.java]: Static import

File: aggregate-snapshot/aggregate-snapshot-service/src/test/java/uk/gov/justice/services/core/aggregate/SnapshotAwareAggregateServiceIT.java [hunk 2]
```diff
@@ -410,7 +413,14 @@ private DynamicallyLoadingClassLoader classLoaderWithGeneratedAggregateLoaded()
...
+                    .with(JsonObjectMetadata
+                            .metadataWithRandomUUID(eventName)
```
Comment (purple52) [inline aggregate-snapshot/aggregate-snapshot-service/src/test/java/uk/gov/justice/services/core/aggregate/SnapshotAwareAggregateServiceIT.java]: Static import

File: aggregate-snapshot/aggregate-snapshot-service/src/test/java/uk/gov/justice/services/core/aggregate/SnapshotAwareAggregateServiceIT.java [hunk 3]
```diff
@@ -419,8 +429,18 @@ private DynamicallyLoadingClassLoader classLoaderWithGeneratedAggregateLoaded()
 
     private <T extends Aggregate> Stream<JsonEnvelope> createEventStreamAndApply(long count, String eventName, T aggregate) {
         List<Object> envelopes = new LinkedList<>();
+
         for (int i = 1; i <= count; i++) {
-            JsonEnvelope envelope = envelope().with(metadataWithRandomUUID(eventName).withStreamId(STREAM_ID)).withPayloadOf("value", "name").build();
+
+            JsonEnvelope envelope = DefaultJsonEnvelope
+                    .envelope()
```
Comment (purple52) [inline aggregate-snapshot/aggregate-snapshot-service/src/test/java/uk/gov/justice/services/core/aggregate/SnapshotAwareAggregateServiceIT.java]: Static import

File: aggregate-snapshot/aggregate-snapshot-service/src/test/java/uk/gov/justice/services/core/aggregate/SnapshotAwareAggregateServiceIT.java [hunk 4]
```diff
@@ -419,8 +429,18 @@ private DynamicallyLoadingClassLoader classLoaderWithGeneratedAggregateLoaded()
...
+                    .with(JsonObjectMetadata
+                            .metadataWithRandomUUID(eventName)
```
Comment (purple52) [inline aggregate-snapshot/aggregate-snapshot-service/src/test/java/uk/gov/justice/services/core/aggregate/SnapshotAwareAggregateServiceIT.java]: Static import

File: aggregate/src/test/java/uk/gov/justice/services/core/aggregate/DefaultAggregateServiceIT.java [hunk 5]
```diff
@@ -207,7 +211,10 @@ private void initDatabase() throws Exception {
     private Stream<JsonEnvelope> envelopes(final int numberOfEnvelopes, String eventName) {
         List<JsonEnvelope> envelopes = new LinkedList<>();
         for (int i = 1; i <= numberOfEnvelopes; i++) {
-            envelopes.add(envelope().with(metadataWithRandomUUID(eventName).withStreamId(STREAM_ID)).withPayloadOf("value", "name").build());
+            envelopes.add(envelope()
+                    .with(metadataWithRandomUUID(eventName).createdAt(clock.now())
```
Comment (purple52) [inline aggregate/src/test/java/uk/gov/justice/services/core/aggregate/DefaultAggregateServiceIT.java]: Split createdAt onto its own line

File: core/src/main/java/uk/gov/justice/services/core/enveloper/Enveloper.java [hunk 6]
```diff
@@ -37,15 +40,15 @@
...
-    private final ObjectToJsonValueConverter objectToJsonValueConverter;
+    Clock clock;
 
-    //for CDI proxying only
-    protected Enveloper() {
-        this.objectToJsonValueConverter = null;
-    }
+    ObjectToJsonValueConverter objectToJsonValueConverter;
+
+    public Enveloper() {}
```
Comment (purple52) [inline core/src/main/java/uk/gov/justice/services/core/enveloper/Enveloper.java]: Why do we have a public default constructor now? I wonder if there is a better way of supporting the EnveloperFactory

File: core/src/test/java/uk/gov/justice/services/core/it/AllEventsHandlerIT.java [hunk 7]
```diff
@@ -11,6 +11,8 @@
 
 import uk.gov.justice.services.common.converter.ObjectToJsonValueConverter;
 import uk.gov.justice.services.common.converter.StringToJsonObjectConverter;
+import uk.gov.justice.services.common.util.Clock;
```
Comment (purple52) [inline core/src/test/java/uk/gov/justice/services/core/it/AllEventsHandlerIT.java]: Is this import used?

File: event-sourcing/event-repository/event-repository-jdbc/src/main/java/uk/gov/justice/services/eventsourcing/repository/jdbc/eventlog/EventLogConverter.java [hunk 8]
```diff
@@ -52,7 +49,7 @@ public EventLog createEventLog(final JsonEnvelope envelope, final UUID streamId,
                 eventMetadata.name(),
                 envelope.metadata().asJsonObject().toString(),
                 extractPayloadAsString(envelope),
-                clock.now());
+                eventMetadata.created().get());
```
Comment (purple52) [inline event-sourcing/event-repository/event-repository-jdbc/src/main/java/uk/gov/justice/services/eventsourcing/repository/jdbc/eventlog/EventLogConverter.java]: We should check there is a value present and throw an exception with an explanation if it is not there.

File: event-sourcing/event-repository/event-repository-jdbc/src/test/java/uk/gov/justice/services/eventsourcing/repository/jdbc/eventlog/EventLogConverterTest.java [hunk 9]
```diff
@@ -90,6 +93,7 @@ private JsonEnvelope createTestEnvelope() throws IOException {
         final Metadata metadata = metadataFrom(Json.createObjectBuilder()
                 .add("id", ID.toString())
                 .add("name", NAME)
+                .add(DATE_CREATED, ZonedDateTimes.toString(clock.now()))
```
Comment (purple52) [inline event-sourcing/event-repository/event-repository-jdbc/src/test/java/uk/gov/justice/services/eventsourcing/repository/jdbc/eventlog/EventLogConverterTest.java]: Looks like we're not testing metadata here - we should add tests to cover this. Probably need to use the StoppedClock for that.

File: event-sourcing/event-repository/event-repository-jdbc/src/test/java/uk/gov/justice/services/eventsourcing/repository/jdbc/eventlog/EventLogJdbcRepositoryIT.java [hunk 10]
```diff
@@ -29,7 +31,7 @@
     private static final String PAYLOAD_JSON = "{\"field\": \"Value\"}";
     private static final String METADATA_JSON = "{\"field\": \"Value\"}";
     private static final String LIQUIBASE_EVENT_STORE_DB_CHANGELOG_XML = "liquibase/event-store-db-changelog.xml";
-    private final static ZonedDateTime TIMESTAMP = now();
```
Comment (purple52) [inline event-sourcing/event-repository/event-repository-jdbc/src/test/java/uk/gov/justice/services/eventsourcing/repository/jdbc/eventlog/EventLogJdbcRepositoryIT.java]: I think there is now an unused static import for now()

File: generators/rest-client-core/src/test/java/uk/gov/justice/services/clients/core/RestClientProcessorBuilder.java [hunk 11]
```diff
@@ -25,7 +30,10 @@ public RestClientProcessor build() {
         final RestClientProcessor restClientProcessor = new RestClientProcessor();
         restClientProcessor.stringToJsonObjectConverter = new StringToJsonObjectConverter();
         restClientProcessor.jsonObjectEnvelopeConverter = new JsonObjectEnvelopeConverter();
-        restClientProcessor.enveloper = new Enveloper(null);
+        final ObjectMapper objectMapper = new ObjectMapperProducer().objectMapper();
+        final ObjectToJsonValueConverter converter = new ObjectToJsonValueConverter(objectMapper);
+        final Enveloper enveloper =  new Enveloper(new UtcClock(), converter);
```
Comment (purple52) [inline generators/rest-client-core/src/test/java/uk/gov/justice/services/clients/core/RestClientProcessorBuilder.java]: Isn't this what the EnveloperFactory is for? Creating real envelopers for tests?

File: messaging/messaging-core/src/main/java/uk/gov/justice/services/messaging/JsonObjectMetadata.java [hunk 12]
```diff
@@ -1,5 +1,8 @@
 package uk.gov.justice.services.messaging;
 
+import uk.gov.justice.services.common.converter.ZonedDateTimes;
+import uk.gov.justice.services.common.util.Clock;
```
Comment (purple52) [inline messaging/messaging-core/src/main/java/uk/gov/justice/services/messaging/JsonObjectMetadata.java]: Unused import? I hope...

File: messaging/messaging-core/src/main/java/uk/gov/justice/services/messaging/Metadata.java [hunk 13]
```diff
@@ -76,4 +77,5 @@
      */
     JsonObject asJsonObject();
 
+    Optional<ZonedDateTime> created();
```
Comment (purple52) [inline messaging/messaging-core/src/main/java/uk/gov/justice/services/messaging/Metadata.java]: Do we want this to be consistently named - I think we called the builder method createdAt. Worth a quick team discussion before it is set in stone.

File: aggregate-snapshot/aggregate-snapshot-service/src/test/java/uk/gov/justice/services/core/aggregate/SnapshotAwareAggregateServiceIT.java (hunk 2 above)
Comment (skiddykong) [inline aggregate-snapshot/aggregate-snapshot-service/src/test/java/uk/gov/justice/services/core/aggregate/SnapshotAwareAggregateServiceIT.java]: Done

File: aggregate/src/test/java/uk/gov/justice/services/core/aggregate/DefaultAggregateServiceIT.java (hunk 5 above)
Comment (skiddykong) [inline aggregate/src/test/java/uk/gov/justice/services/core/aggregate/DefaultAggregateServiceIT.java]: done

File: core/src/main/java/uk/gov/justice/services/core/enveloper/Enveloper.java (hunk 6 above)
Comment (skiddykong) [inline core/src/main/java/uk/gov/justice/services/core/enveloper/Enveloper.java]: no default constructor results in: 

SEVERE - CDI Beans module deployment failed
org.apache.webbeans.exception.WebBeansDeploymentException: javax.enterprise.inject.UnproxyableResolutionException: WebBeans with api type with normal scope must be proxyable.
uk.gov.justice.services.core.enveloper.Enveloper has no explicit no-arg constructor!A public or protected constructor without args is required!
	at org.apache.webbeans.config.BeansDeployer.deploy(BeansDeployer.java:334)
	at org.apache.openejb.cdi.OpenEJBLifecycle.startApplication(OpenEJBLifecycle.java:196)
	at org.apache.openejb.cdi.ThreadSingletonServiceImpl.initialize(ThreadSingletonServiceImpl.java:193)
	at org.apache.openejb.cdi.CdiBuilder.build(CdiBuilder.java:41)
	at org.apache.openejb.assembler.classic.Assembler.createApplication(Assembler.java:895)
	at org.apache.openejb.assembler.classic.Assembler.createApplication(Assembler.java:694)
	at org.apache.openejb.assembler.classic.Assembler.createApplication(Assembler.java:690)
	at org.apache.openejb.testing.ApplicationComposers.deployApp(ApplicationComposers.java:721)
	at org.apache.openejb.testing.ApplicationComposers.before(ApplicationComposers.java:383)
	at org.apache.openejb.testing.ApplicationComposers.evaluate(ApplicationComposers.java:1060)
	at org.apache.openejb.junit.DeployApplication.evaluate(DeployApplication.java:40)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:367)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:274)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:161)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:290)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:242)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:121)
Caused by: javax.enterprise.inject.UnproxyableResolutionException: WebBeans with api type with normal scope must be proxyable.
uk.gov.justice.services.core.enveloper.Enveloper has no explicit no-arg constructor!A public or protected constructor without args is required!
	at org.apache.webbeans.util.InjectionExceptionUtil.createUnproxyableResolutionException(InjectionExceptionUtil.java:46)
	at org.apache.webbeans.util.InjectionExceptionUtil.throwUnproxyableResolutionException(InjectionExceptionUtil.java:41)
	at org.apache.webbeans.config.DeploymentValidationService.validateProxyable(DeploymentValidationService.java:143)
	at org.apache.webbeans.component.creation.ManagedBeanBuilder.getBean(ManagedBeanBuilder.java:70)
	at org.apache.webbeans.config.BeansDeployer.defineManagedBean(BeansDeployer.java:1629)
	at org.apache.webbeans.config.BeansDeployer.deploySingleAnnotatedType(BeansDeployer.java:1255)
	at org.apache.webbeans.config.BeansDeployer.deployFromBeanAttributes(BeansDeployer.java:1211)
	at org.apache.webbeans.config.BeansDeployer.deploy(BeansDeployer.java:292)
	... 26 more

Comment (skiddykong) [inline core/src/main/java/uk/gov/justice/services/core/enveloper/Enveloper.java]: Caused by: javax.enterprise.inject.UnproxyableResolutionException: WebBeans with api type with normal scope must be proxyable.
uk.gov.justice.services.core.enveloper.Enveloper has no explicit no-arg constructor!A public or protected constructor without args is required!

File: core/src/test/java/uk/gov/justice/services/core/it/AllEventsHandlerIT.java (hunk 7 above)
Comment (skiddykong) [inline core/src/test/java/uk/gov/justice/services/core/it/AllEventsHandlerIT.java]: removed

File: event-sourcing/event-repository/event-repository-jdbc/src/main/java/uk/gov/justice/services/eventsourcing/repository/jdbc/eventlog/EventLogConverter.java (hunk 8 above)
Comment (skiddykong) [inline event-sourcing/event-repository/event-repository-jdbc/src/main/java/uk/gov/justice/services/eventsourcing/repository/jdbc/eventlog/EventLogConverter.java]: done

File: core/src/main/java/uk/gov/justice/services/core/enveloper/Enveloper.java (hunk 6 above)
Comment (skiddykong) [inline core/src/main/java/uk/gov/justice/services/core/enveloper/Enveloper.java]: Got it, have made protected, compiles fine

File: event-sourcing/event-repository/event-repository-jdbc/src/test/java/uk/gov/justice/services/eventsourcing/repository/jdbc/eventlog/EventLogConverterTest.java (hunk 9 above)
Comment (skiddykong) [inline event-sourcing/event-repository/event-repository-jdbc/src/test/java/uk/gov/justice/services/eventsourcing/repository/jdbc/eventlog/EventLogConverterTest.java]: added test & used StoppedClock

File: event-sourcing/event-repository/event-repository-jdbc/src/test/java/uk/gov/justice/services/eventsourcing/repository/jdbc/eventlog/EventLogJdbcRepositoryIT.java (hunk 10 above)
Comment (skiddykong) [inline event-sourcing/event-repository/event-repository-jdbc/src/test/java/uk/gov/justice/services/eventsourcing/repository/jdbc/eventlog/EventLogJdbcRepositoryIT.java]: done

File: generators/rest-client-core/src/test/java/uk/gov/justice/services/clients/core/RestClientProcessorBuilder.java (hunk 11 above)
Comment (skiddykong) [inline generators/rest-client-core/src/test/java/uk/gov/justice/services/clients/core/RestClientProcessorBuilder.java]: done

File: messaging/messaging-core/src/main/java/uk/gov/justice/services/messaging/JsonObjectMetadata.java (hunk 12 above)
Comment (skiddykong) [inline messaging/messaging-core/src/main/java/uk/gov/justice/services/messaging/JsonObjectMetadata.java]: removed

File: messaging/messaging-core/src/main/java/uk/gov/justice/services/messaging/Metadata.java (hunk 13 above)
Comment (skiddykong) [inline messaging/messaging-core/src/main/java/uk/gov/justice/services/messaging/Metadata.java]: in the database table the column is called date_created

Comment (skiddykong) [inline messaging/messaging-core/src/main/java/uk/gov/justice/services/messaging/Metadata.java]: changed to createdAt

**Pull request merged**